lib/Runtime/Debug/TTActionEvents.cpp

```diff
@@ -0,0 +1,986 @@
+            cpInfo->SourceCode.Contents = alloc.SlabAllocateArray<wchar>(cpInfo->SourceCode.Length + 1);
+
+            UtilSupport::TTAutoString docId;
+            docId.Append(cpInfo->DocumentID);
+            docId.Append(_u("ld"));
+
+            JsSupport::ReadCodeFromFile(threadContext->TTDStreamFunctions, cpInfo->SrcDir.Contents, docId.GetStrValue(), cpInfo->SourceUri.Contents, cpInfo->SourceCode.Contents, cpInfo->SourceCode.Length);
+        }
+
+#if ENABLE_TTD_INTERNAL_DIAGNOSTICS
+        int64 JsRTCallFunctionAction_GetLastNestedEventTime(const EventLogEntry* evt)
+        {
+            const JsRTCallFunctionAction* cfAction = GetInlineEventDataAs<JsRTCallFunctionAction, EventKind::CallExistingFunctionActionTag>(evt);
+
+            return cfAction->AdditionalInfo->LastNestedEvent;
+        }
+
+        void JsRTCallFunctionAction_ProcessDiagInfoPre(EventLogEntry* evt, Js::JavascriptFunction* function, UnlinkableSlabAllocator& alloc)
+        {
+            JsRTCallFunctionAction* cfAction = GetInlineEventDataAs<JsRTCallFunctionAction, EventKind::CallExistingFunctionActionTag>(evt);
+
+            Js::JavascriptString* displayName = function->GetDisplayName();
+            alloc.CopyStringIntoWLength(displayName->GetSz(), displayName->GetLength(), cfAction->AdditionalInfo->FunctionName);
+        }
+
+        void JsRTCallFunctionAction_ProcessDiagInfoPost(EventLogEntry* evt, double wallTime, int64 lastNestedEvent)
+        {
+            JsRTCallFunctionAction* cfAction = GetInlineEventDataAs<JsRTCallFunctionAction, EventKind::CallExistingFunctionActionTag>(evt);
+
+            cfAction->AdditionalInfo->EndTime = wallTime;
+            cfAction->AdditionalInfo->LastNestedEvent = lastNestedEvent;
+        }
+#endif
+
+        void JsRTCallFunctionAction_ProcessArgs(EventLogEntry* evt, int32 rootDepth, int64 callEventTime, Js::JavascriptFunction* function, uint32 argc, Js::Var* argv, double wallTime, int64 hostCallbackId, int64 topLevelCallbackEventTime, UnlinkableSlabAllocator& alloc)
+        {
+            JsRTCallFunctionAction* cfAction = GetInlineEventDataAs<JsRTCallFunctionAction, EventKind::CallExistingFunctionActionTag>(evt);
+            cfAction->AdditionalInfo = alloc.SlabAllocateStruct<JsRTCallFunctionAction_AdditionalInfo>();
+
+            cfAction->CallbackDepth = rootDepth;
+            cfAction->ArgCount = argc + 1;
+
+            static_assert(sizeof(TTDVar) == sizeof(Js::Var), "These need to be the same size (and have same bit layout) for this to work!");
+
+            cfAction->ArgArray = alloc.SlabAllocateArray<TTDVar>(cfAction->ArgCount);
+            cfAction->ArgArray[0] = TTD_CONVERT_JSVAR_TO_TTDVAR(function);
+            js_memcpy_s(cfAction->ArgArray + 1, (cfAction->ArgCount -1) * sizeof(TTDVar), argv, argc * sizeof(Js::Var));
+
+            cfAction->AdditionalInfo->BeginTime = wallTime;
+            cfAction->AdditionalInfo->EndTime = -1.0;
```

Comment:
All these special magic numbers should be a #define so that it is easy to understand what they mean.